File: apps/admin-ui/src/api/quality.ts
/**
 * Quality-dashboard SDK — backed by ``/v1/quality`` (RT-5 / RT-ADR-26).
 *
 * Returns the **raw** payload (no ``{ success, data, error }`` envelope),
 * consistent with the control-plane ``api/quality.py`` (which mirrors
 * ``api/eval_runs.py``). So these calls go through ``apiClient`` directly and
 * read ``response.data``, NOT through an envelope unwrap.
 *
 * Scope: home-tenant only — the server derives the tenant from the request
 * (RLS), so this SDK threads no ``tenant_id``.
 */
import { apiClient } from "./client";

/** One LLM-judge verdict for a sampled production run. ``overall`` /
 *  ``dimensions`` are a subjective 1-5 rubric score, not ground truth. */
export interface QualityScore {
  id: number;
  agent_name: string;
  agent_version: string;
  /** For drill-down: link to the run / conversation in run_detail. */
  run_id: string;
  thread_id: string;
  overall: number;
  dimensions: Record<string, number>;
  rationale: string;
  judge_model: string;
  observed_at: string | null;
}

/** One raised drift alert (recent mean dropped below baseline). */
export interface QualityDriftAlert {
  id: number;
  agent_name: string;
  recent_mean: number;
  baseline_mean: number;
  /** Relative drop ``(baseline - recent) / baseline``. */
  drift_pct: number;
  recent_count: number;
  baseline_count: number;
  detected_at: string | null;
}

export interface QualityScoreList {
  items: QualityScore[];
}

export interface QualityDriftAlertList {
  items: QualityDriftAlert[];
}

export interface ListQualityScoresParams {
  agentName?: string;
  windowH?: number;
  limit?: number;
}

export interface ListQualityDriftAlertsParams {
  agentName?: string;
  windowH?: number;
  limit?: number;
}

/** GET /v1/quality/scores — the per-agent score series (newest first). */
export async function listQualityScores(
  params: ListQualityScoresParams = {},
): Promise<QualityScoreList> {
  const { agentName, windowH, limit } = params;
  const response = await apiClient.get<QualityScoreList>("/v1/quality/scores", {
    params: { agent_name: agentName, window_h: windowH, limit },
  });
  return response.data;
}

/** GET /v1/quality/drift-alerts — raised drift alerts (newest first). */
export async function listQualityDriftAlerts(
  params: ListQualityDriftAlertsParams = {},
): Promise<QualityDriftAlertList> {
  const { agentName, windowH, limit } = params;
  const response = await apiClient.get<QualityDriftAlertList>(
    "/v1/quality/drift-alerts",
    {
      params: { agent_name: agentName, window_h: windowH, limit },
    },
  );
  return response.data;
}

File: services/control-plane/src/control_plane/api/quality.py
"""``/v1/quality`` — Stream RT-5 (RT-ADR-26) production-quality dashboard reads.

Home-tenant, read-only surface over the ``quality_score`` time-series and the
``quality_drift_alert`` history the RT-5 workers fill:

- ``GET /v1/quality/scores`` — the per-agent score series (trend chart) and the
  low-score rows a caller drills into (each carries ``run_id`` / ``thread_id``
  so the UI can link to ``run_detail``).
- ``GET /v1/quality/drift-alerts`` — the raised drift alerts.

Reads never cross tenants: the caller's tenant (RLS GUC set by the request
middleware) plus the store's explicit ``tenant_id`` filter both enforce it.
Returns the **raw** payload (no ``{success, data, error}`` envelope), matching
``api/eval_runs.py`` and the other operator read surfaces.
"""

from __future__ import annotations

from datetime import UTC, datetime, timedelta
from typing import Annotated, Any
from uuid import UUID

from fastapi import APIRouter, Depends, Query, Request
from fastapi.responses import JSONResponse

from expert_work.persistence import QualityDriftAlertStore, QualityScoreStore
from expert_work.protocol import QualityDriftAlertRecord, QualityScoreRecord


def _score_dict(record: QualityScoreRecord) -> dict[str, Any]:
    return {
        "id": record.id,
        "agent_name": record.agent_name,
        "agent_version": record.agent_version,
        "run_id": str(record.run_id),
        "thread_id": str(record.thread_id),
        "overall": record.overall,
        "dimensions": record.dimensions,
        "rationale": record.rationale,
        "judge_model": record.judge_model,
        "observed_at": record.observed_at.isoformat() if record.observed_at else None,
    }


def _alert_dict(record: QualityDriftAlertRecord) -> dict[str, Any]:
    return {
        "id": record.id,
        "agent_name": record.agent_name,
        "recent_mean": record.recent_mean,
        "baseline_mean": record.baseline_mean,
        "drift_pct": record.drift_pct,
        "recent_count": record.recent_count,
        "baseline_count": record.baseline_count,
        "detected_at": record.detected_at.isoformat() if record.detected_at else None,
    }


def _get_score_store(request: Request) -> QualityScoreStore:
    return request.app.state.quality_score_store  # type: ignore[no-any-return]


def _get_alert_store(request: Request) -> QualityDriftAlertStore:
    return request.app.state.quality_drift_alert_store  # type: ignore[no-any-return]


def build_quality_router() -> APIRouter:
    """Read the per-agent quality series + drift alerts (home-tenant)."""
    router = APIRouter(prefix="/v1/quality", tags=["quality"])

    @router.get("/scores", response_model=None)
    async def list_scores(
        request: Request,
        store: Annotated[QualityScoreStore, Depends(_get_score_store)],
        agent_name: Annotated[str | None, Query(max_length=200)] = None,
        window_h: Annotated[int, Query(ge=1, le=8760)] = 168,
        limit: Annotated[int, Query(ge=1, le=1000)] = 500,
    ) -> JSONResponse:
        tenant_id: UUID = request.state.tenant_id
        since = datetime.now(tz=UTC) - timedelta(hours=window_h)
        rows = await store.list_scores(
            tenant_id=tenant_id, agent_name=agent_name, since=since, limit=limit
        )
        return JSONResponse(content={"items": [_score_dict(r) for r in rows]})

    @router.get("/drift-alerts", response_model=None)
    async def list_drift_alerts(
        request: Request,
        store: Annotated[QualityDriftAlertStore, Depends(_get_alert_store)],
        agent_name: Annotated[str | None, Query(max_length=200)] = None,
        window_h: Annotated[int, Query(ge=1, le=8760)] = 720,
        limit: Annotated[int, Query(ge=1, le=500)] = 100,
    ) -> JSONResponse:
        tenant_id: UUID = request.state.tenant_id
        since = datetime.now(tz=UTC) - timedelta(hours=window_h)
        rows = await store.list_alerts(
            tenant_id=tenant_id, agent_name=agent_name, since=since, limit=limit
        )
        return JSONResponse(content={"items": [_alert_dict(r) for r in rows]})

    return router


__all__ = ["build_quality_router"]
